### PR TITLE
xen 4.12 separate stubdom build

### PR DESCRIPTION
**Pull request by fepitre**

### CODE REVIEW

File: xen-hvm-stubdom-legacy.spec.in [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+
+tar xf %{SOURCE35}
+cp -a stubdom-dhcp/* tools/qemu-xen-traditional/
+patch -d tools/qemu-xen-traditional -p4 < stubdom-dhcp/lwip-dhcp-qemu-glue.patch
+
+
+%build
+
+%define efi_flags EFI_VENDOR=qubes
+mkdir -p dist/install/boot/efi/efi/qubes
```
Comment (marmarek) [inline xen-hvm-stubdom-legacy.spec.in]: Is this really needed? I don't think you build (or need) xen.efi here.

File: xen-hvm-stubdom-legacy.spec.in [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+
+export XEN_VENDORVERSION="-%{release}"
+export EXTRA_CFLAGS_XEN_TOOLS="$RPM_OPT_FLAGS"
+export EXTRA_CFLAGS_QEMU_TRADITIONAL="$RPM_OPT_FLAGS"
+export EXTRA_CFLAGS_QEMU_XEN="$RPM_OPT_FLAGS"
+export PYTHON="/usr/bin/python2"
```
Comment (marmarek) [inline xen-hvm-stubdom-legacy.spec.in]: Have you tried with python3?

File: xen-hvm-stubdom-legacy.spec.in [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+    --disable-rombios \
+    --disable-ipxe \
+    --disable-systemd \
+    --disable-9pfs
+
+export RPM_OPT_FLAGS_RED=`echo $RPM_OPT_FLAGS | sed -e 's/-m64//g' -e 's/--param=ssp-buffer-size=4//g' -e's/-fstack-protector-strong//'`
+export EXTRA_CFLAGS_XEN_TOOLS="$RPM_OPT_FLAGS_RED"
+
+# tools
+make %{?_smp_mflags} prefix=/usr tools
```
Comment (marmarek) [inline xen-hvm-stubdom-legacy.spec.in]: Have you tried without this step?

File: xen-hvm-stubdom-legacy.spec.in [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+# modified vs upstream
+Patch916: patch-stubdom-linux-v3-0017-libxl-require-qemu-in-dom0-even-if-stubdomain-is-in-use.patch
+
+# MSI fixes
+Patch920: patch-stubdom-msi-0001-libxl-do-not-attach-xen-pciback-to-HVM-domain-if-stu.patch
+Patch921: patch-stubdom-msi-0002-libxl-attach-PCI-device-to-qemu-only-after-setting-p.patch
+Patch922: patch-stubdom-msi-0003-libxl-don-t-try-to-manipulate-json-config-for-stubdo.patch
+Patch923: patch-stubdom-msi-0004-xen-x86-Allow-stubdom-access-to-irq-created-for-msi.patch
+Patch924: patch-stubdom-msi-0005-xen-x86-add-PHYSDEVOP_msi_set_enable.patch
+Patch925: patch-stubdom-msi-0006-tools-libxc-add-wrapper-for-PHYSDEVOP_msi_set_enable.patch
```
Comment (marmarek) [inline xen-hvm-stubdom-legacy.spec.in]: Those patches are for xen (and toolstack) side of stubdomain, not needed here.

File: xen-hvm-stubdom-legacy.spec.in [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+Patch908: patch-stubdom-linux-v3-0009-tools-libvchan-notify-server-when-client-is-connected.patch
+Patch909: patch-stubdom-linux-v3-0010-libxl-typo-fix-in-comment.patch
+Patch910: patch-stubdom-linux-v3-0011-libxl-move-xswait-declaration-up-in-libxl_internal.h.patch
+Patch911: patch-stubdom-linux-v3-0012-libxl-use-vchan-for-QMP-access-with-Linux-stubdomain-libxl__ev_qmp_-version.patch
+Patch912: patch-stubdom-linux-v3-0013-libxl-use-vchan-for-QMP-access-with-Linux-stubdomain-non-async-version.patch
+Patch913: patch-stubdom-linux-v3-0014-libxl-add-save-restore-support-for-qemu-xen-in-stubdomain.patch
+Patch914: patch-stubdom-linux-v3-0015-tools-add-missing-libxenvchan-cflags.patch
+Patch915: patch-stubdom-linux-v3-0016-libxl-add-locking-for-libvchan-QMP-connection.patch
+# modified vs upstream
+Patch916: patch-stubdom-linux-v3-0017-libxl-require-qemu-in-dom0-even-if-stubdomain-is-in-use.patch
```
Comment (marmarek) [inline xen-hvm-stubdom-legacy.spec.in]: Those are for toolstack side of stubdomain (and linux one, not legacy), not needed here.

File: xen-hvm-stubdom-legacy.spec.in [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+
+# Security fixes
+
+# Upstreamable patches
+Patch601: patch-xen-libxl-error-write-perm.patch
+#TODO: simplified script instead:
+# Patch604: patch-libxl-Revert-libxl-Remove-redundant-setting-of-phyical-dev.patch
+# Patch605: patch-libxl-allow-PHY-backend-for-files-allocate-loop-devi.patch
+# Patch606: patch-libxl-do-not-call-default-block-script.patch
+Patch607: patch-libxl-do-not-for-backend-on-PCI-remove-when-backend-.patch
```
Comment (marmarek) [inline xen-hvm-stubdom-legacy.spec.in]: This one is for toolstack, not needed here.

File: xen-hvm-stubdom-legacy.spec.in [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+Patch610: patch-stubdom-gmp-compile-fix.patch
+Patch614: patch-0001-libxl-do-not-fail-device-removal-if-backend-domain-i.patch
```
Comment (marmarek) [inline xen-hvm-stubdom-legacy.spec.in]: And this too.

File: xen-hvm-stubdom-legacy.spec.in [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+Patch618: patch-hvmpt04-minios-nomask-bar-addrs.patch
+Patch619: patch-hvmpt05-hide-pio-bars.patch
+Patch620: patch-hvmpt06-fix-msix.patch
+Patch621: patch-libxc-fix-xc_gntshr_munmap-semantic.patch
+Patch622: patch-minios-ignore-close-0.patch
+Patch624: patch-qemu-docs-utf8.patch
+Patch625: patch-minios-remove-net-device-instead-of-preparing-for-re.patch
+Patch627: patch-qemu-keyboard-leds.patch
+Patch628: patch-python-Adjust-xc_physinfo-wrapper-for-updated-virt_c.patch
+Patch634: patch-python-do-not-report-handled-EAGAIN-error.patch
```
Comment (marmarek) [inline xen-hvm-stubdom-legacy.spec.in]: The last two too.

File: xen-hvm-stubdom-legacy.spec.in [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+Patch921: patch-stubdom-msi-0002-libxl-attach-PCI-device-to-qemu-only-after-setting-p.patch
+Patch922: patch-stubdom-msi-0003-libxl-don-t-try-to-manipulate-json-config-for-stubdo.patch
+Patch923: patch-stubdom-msi-0004-xen-x86-Allow-stubdom-access-to-irq-created-for-msi.patch
+Patch924: patch-stubdom-msi-0005-xen-x86-add-PHYSDEVOP_msi_set_enable.patch
+Patch925: patch-stubdom-msi-0006-tools-libxc-add-wrapper-for-PHYSDEVOP_msi_set_enable.patch
+
+# Qubes specific patches
+Patch1001: patch-stubdom-vbd-non-dom0-backend.patch
+Patch1002: patch-xen-no-downloads.patch
+Patch1003: patch-xen-hotplug-external-store.patch
```
Comment (marmarek) [inline xen-hvm-stubdom-legacy.spec.in]: This is not needed for stubdomain build.

File: xen-hvm-stubdom-legacy.spec.in [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+Patch1004: patch-xen-stubdom-qubes-gui.patch
+Patch1005: patch-stubdom-lwip-fix-for-dhcp.patch
+Patch1006: patch-xen-libxl-qubes-minimal-stubdom.patch
+Patch1007: patch-xen-disable-dom0-qemu.patch
```
Comment (marmarek) [inline xen-hvm-stubdom-legacy.spec.in]: This is not needed for stubdomain build.

File: xen-hvm-stubdom-legacy.spec.in [hunk 11]
```diff
@@ -0,0 +1,291 @@
...
+Patch1009: patch-xenconsoled-enable-logging.patch
```
Comment (marmarek) [inline xen-hvm-stubdom-legacy.spec.in]: This is not needed for stubdomain build.

File: xen-hvm-stubdom-legacy.spec.in [hunk 12]
```diff
@@ -0,0 +1,291 @@
...
+#TODO Patch1110: patch-stubdom-linux-libxl-suspend.patch
+Patch1011: patch-xen-hotplug-qubesdb-update.patch
```
Comment (marmarek) [inline xen-hvm-stubdom-legacy.spec.in]: This is not needed for stubdomain build.

File: xen-hvm-stubdom-legacy.spec.in [hunk 13]
```diff
@@ -0,0 +1,291 @@
...
+Patch1012: patch-tools-hotplug-drop-perl-usage-in-locking-mechanism.patch
```
Comment (marmarek) [inline xen-hvm-stubdom-legacy.spec.in]: This is not needed for stubdomain build.

File: xen-hvm-stubdom-legacy.spec.in [hunk 14]
```diff
@@ -0,0 +1,291 @@
...
+Patch1013: patch-stubdom-linux-libxl-use-EHCI-for-providing-tablet-USB-device.patch
```
Comment (marmarek) [inline xen-hvm-stubdom-legacy.spec.in]: This is not needed for legacy stubdomain build.

File: xen-hvm-stubdom-legacy.spec.in [hunk 15]
```diff
@@ -0,0 +1,291 @@
...
+Patch1014: patch-allow-kernelopts-stubdom.patch
```
Comment (marmarek) [inline xen-hvm-stubdom-legacy.spec.in]: This is not needed for stubdomain build (toolstack patch).

File: xen-hvm-stubdom-legacy.spec.in [hunk 16]
```diff
@@ -0,0 +1,291 @@
...
+Patch1015: patch-libxl-readonly-disk-scsi.patch
+Patch1016: patch-tools-xenconsole-replace-ESC-char-on-xenconsole-outp.patch
```
Comment (marmarek) [inline xen-hvm-stubdom-legacy.spec.in]: And those two too.

File: xen-hvm-stubdom-legacy.spec.in [hunk 17]
```diff
@@ -0,0 +1,291 @@
...
+
+Patch1020: patch-stubdom-linux-config-qubes-gui.patch
+Patch1021: patch-stubdom-linux-libxl-do-not-force-qdisk-backend-for-cdrom.patch
+Patch1022: patch-xen-acpi-slic-support.patch
```
Comment (marmarek) [inline xen-hvm-stubdom-legacy.spec.in]: Those 3 are toolstack patches, not needed here.

File: xen-hvm-stubdom-legacy.spec.in [hunk 18]
```diff
@@ -0,0 +1,291 @@
...
+
+Patch1100: patch-fix-errors-on-ambiguous-python-shebang.patch
```
Comment (marmarek) [inline xen-hvm-stubdom-legacy.spec.in]: Is rpm complaining without this? The package shouldn't contain any python files...

File: xen.spec.in [hunk 19]
```diff
@@ -74,23 +73,17 @@ License: GPLv2+ and LGPLv2+ and BSD
...
 Source10: lwip-1.3.0.tar.gz
 Source11: newlib-1.16.0.tar.gz
 Source12: zlib-1.2.3.tar.gz
 Source13: pciutils-2.2.9.tar.bz2
 Source14: grub-0.97.tar.gz
-Source15: gmp-4.3.2.tar.bz2
-Source16: polarssl-1.1.4-gpl.tgz
-Source18: tpm_emulator-0.7.4.tar.gz
-Source32: xen.modules-load.conf
+Source15: polarssl-1.1.4-gpl.tgz
```
Comment (marmarek) [inline xen.spec.in]: I think none of those tarballs should be needed here anymore.

File: xen.spec.in [hunk 20]
```diff
@@ -121,86 +114,26 @@ Patch601: patch-xen-libxl-error-write-perm.patch
...
 Patch1011: patch-xen-hotplug-qubesdb-update.patch
 Patch1012: patch-tools-hotplug-drop-perl-usage-in-locking-mechanism.patch
-Patch1013: patch-stubdom-linux-libxl-use-EHCI-for-providing-tablet-USB-device.patch
-Patch1014: patch-allow-kernelopts-stubdom.patch
 Patch1015: patch-libxl-readonly-disk-scsi.patch
 Patch1016: patch-tools-xenconsole-replace-ESC-char-on-xenconsole-outp.patch
-
-Patch1020: patch-stubdom-linux-config-qubes-gui.patch
-Patch1021: patch-stubdom-linux-libxl-do-not-force-qdisk-backend-for-cdrom.patch
 Patch1022: patch-xen-acpi-slic-support.patch
```
Comment (marmarek) [inline xen.spec.in]: to the whole above list: adjust according to comments in the other spec

File: xen.spec.in [hunk 21]
```diff
@@ -472,30 +405,10 @@ Just a few xenstore-* tools and Xen hotplug scripts needed by Qubes VMs
...
-
-tar xf %{SOURCE34}
-cp -a core-vchan-xen/vchan tools/
-make -C tools/vchan -f Makefile.stubdom clean
-sed -e 's/ioemu-qemu-xen/qemu-xen-traditional/g' tools/qubes-gui/gui-agent-qemu/qemu-glue.patch | patch -p1
-
-tar xf %{SOURCE35}
-cp -a stubdom-dhcp/* tools/qemu-xen-traditional/
-patch -d tools/qemu-xen-traditional -p4 < stubdom-dhcp/lwip-dhcp-qemu-glue.patch
+cp -v %{SOURCE10} %{SOURCE11} %{SOURCE12} %{SOURCE13} %{SOURCE14} %{SOURCE15} stubdom
```
Comment (marmarek) [inline xen.spec.in]: This all (hopefully) should not be needed anymore.

File: xen.spec.in [hunk 22]
```diff
@@ -556,8 +468,6 @@ autoreconf
     --libexecdir=%{_libexecdir} \
     --with-system-seabios=%{seabiosloc} \
     --with-system-ipxe=/usr/share/ipxe/8086100e.rom \
-    --enable-vtpm-stubdom \
-    --enable-vtpmmgr-stubdom \
```
Comment (marmarek) [inline xen.spec.in]: I think there is also `--disable-stubdom` (see main configure help).

And you might also need to adjust %files section in this spec.

File: xen-hvm-stubdom-legacy.spec.in (hunk 3 above)
Comment (fepitre) [reply xen-hvm-stubdom-legacy.spec.in]: Yes it's not working due to xentoolcore_internal.h.

File: xen-hvm-stubdom-legacy.spec.in (hunk 2 above)
Comment (fepitre) [reply xen-hvm-stubdom-legacy.spec.in]: Yes with new newer commits on 'upstream' branch.

File: xen-hvm-stubdom-legacy.spec.in (hunk 10 above)
Comment (marmarek) [reply xen-hvm-stubdom-legacy.spec.in]: I mean, the 1007 one.

Comment (marmarek) [reply xen-hvm-stubdom-legacy.spec.in]: and 1006 too, but not 1004 and 1005

File: xen.spec.in (hunk 22 above)
Comment (fepitre) [reply xen.spec.in]: It's already handled. I enforced ```%define build_stubdom 0```.

File: xen-hvm-stubdom-legacy.spec.in [hunk 23]
```diff
@@ -0,0 +1,204 @@
...
+Source18: tpm_emulator-0.7.4.tar.gz
+Source32: xen.modules-load.conf
+
+# Qubes components for stubdom
+Source33: @gui-agent-xen-hvm-stubdom@
+Source34: @core-vchan-xen@
+Source35: @stubdom-dhcp@
+Source36: @gui-common@
+
+Patch601: patch-xen-libxl-error-write-perm.patch
```
Comment (marmarek) [inline xen-hvm-stubdom-legacy.spec.in]: This one looks toolstack related, not stubdom.

**Comment by fepitre** [comment]
@marmarek: I updated the new repo too.

**Pull request merged**